mail_spam/models/reverend_thomas.py

```diff
@@ -0,0 +1,152 @@
+    SPAM = 'spam'
+
+    name = fields.Char(
+        required=True,
+    )
+    database = fields.Binary(
+        attachment=True,
+        readonly=True,
+    )
+    client = fields.Binary(
+        compute='_compute_client',
+    )
+
+    @api.multi
+    @api.depends('database')
+    def _compute_client(self):
+        """Compute the client, optionally loading the stored database."""
+        for record in self.filtered(lambda r: r.database):
+            record.client = self._get_client()
+            with BytesIO(record.database.decode('base64')) as fp:
+                record.client.load_handler(fp)
+
+    @api.model
+    def create(self, vals):
+        """Add a new database into vals if one isn't provided."""
+        record = super(ReverendThomas, self).create(vals)
+        if not vals.get('database'):
+            client = self._get_client()
+            with BytesIO() as fp:
+                client.save_handler(fp)
+                record.database = fp.getvalue().encode('base64')
+        return record
+
+    @api.multi
+    def check(self, message):
+        """Check the parsed message and return the results.
+
+        If partners is provided, it will use the Pyzor servers that are mapped
+        to their company. Otherwise it will use the current user's company.
+
+        Args:
+            message (MailMessage): Message singleton to check as SPAM.
+
+        Returns:
+            dict: Mapping with keys `spam`, `ham`, and `ratio`.
+        """
+
+        message.ensure_one()
+
+        averages = {}
+        output = defaultdict(dict)
+
+        for record in self:
+            for result in record.client.guess(self._parse_message(message)):
+                key, score = result
+                averages[key] = averages.get(key, 0) + score
+                output[record.id][key] = score
+
+        averages['ratio'] = averages[self.HAM] / averages[self.SPAM]
+        output.update(averages)
+        return output
+
+    @api.multi
+    def train_spam(self, messages):
+        """Report a message as SPAM.
+
+        Args:
+            messages (MailMessage): ``mail.message`` recordset to mark as SPAM.
+        """
+        self._train(messages, self.SPAM)
+
+    @api.multi
+    def train_ham(self, messages):
+        """Report a message as SPAM.
+
+        Args:
+            messages (MailMessage): ``mail.message`` recordset to mark as HAM.
+        """
+        self._train(messages, self.HAM)
+
+    @api.multi
+    def _train(self, messages, spam_or_ham):
+        """Train the Bayesian filter.
+
+        Args:
+            messages (MailMessage): ``mail.message`` recordset to train with.
+            spam_or_ham (str): Value to train as (either self.SPAM or self.HAM)
+        """
+        for record in self:
+            for message in messages:
+                message_string = self._parse_message(message)
+                record.client.train(message_string, spam_or_ham)
+        self._save_db()
+
+    @api.model_cr_context
```

Comment:
Weird decorator. Why not model or multi? .. or nothing?

Reply:
This is new in v10 and is meant to decorate something that can be called from both a multi and a model api without issues.

The purpose of these methods is to be as useful as possible. Using either `multi`, or `model` limits that usefulness by requiring the method be called from within that type of decorator.

That said, the best way to go about this is to circumvent Odoo model api entirely and use a combination of class and static methods here.